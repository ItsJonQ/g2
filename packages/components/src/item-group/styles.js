import { css, ui } from '@wp-g2/styles';

// @todo: Maybe abstract to a dedicated UnstyledButton component.
export const unstyledButton = css`
	appearance: none;
	background: none;
	border: 1px solid transparent;
	cursor: pointer;
	text-align: left;

	&:hover {
		color: ${ui.get('colorAdmin')};
	}

	&:focus {
		background-color: transparent;
		border-color: ${ui.get('colorAdmin')};
		color: ${ui.get('colorAdmin')};
		outline: 3px solid transparent;
	}
`;

export const item = css`
	display: block;
	width: 100%;
`;

export const bordered = css`
	${ui.border.all}
`;

export const separated = css`
	> *:not(marquee) {
		${ui.border.bottom}
	}

	> *:last-child:not(:focus) {
		border-bottom-color: transparent;
	}
`;

const borderRadius = ui.get('controlBorderRadius');

export const spacedAround = css`
	${ui.borderRadius(borderRadius)}
`;

export const rounded = css`
	${ui.borderRadius(borderRadius)}

	> *:first-child {
		border-top-left-radius: ${borderRadius};
		border-top-right-radius: ${borderRadius};
	}

	> *:last-child {
		border-bottom-left-radius: ${borderRadius};
		border-bottom-right-radius: ${borderRadius};
	}
`;

const baseFontHeight = `calc(${ui.get('fontSize')} * ${ui.get(
	'fontLineHeightBase',
)})`;

/*
 * Math:
 * - Use the desired height as the base value
 * - Subtract the computed height of (default) text
 * - Subtract the effects of border
 * - Divide the calculated number by 2, in order to get an individual top/bottom padding
 */
const paddingY = `calc((${ui.get(
	'controlHeight',
)} - ${baseFontHeight} - 2px) / 2)`;
const paddingYSmall = `calc((${ui.get(
	'controlHeightSmall',
)} - ${baseFontHeight} - 2px) / 2)`;
const paddingYLarge = `calc((${ui.get(
	'controlHeightLarge',
)} - ${baseFontHeight} - 2px) / 2)`;

export const itemSizes = {
	small: css`
		${ui.padding.y(paddingYSmall)}
		${ui.padding.x(ui.get('controlPaddingXSmall'))}
	`,
	medium: css`
		${ui.padding.y(paddingY)}
		${ui.padding.x(ui.get('controlPaddingX'))}
	`,
	large: css`
		${ui.padding.y(paddingYLarge)}
		${ui.padding.x(ui.get('controlPaddingXLarge'))}
	`,
};
