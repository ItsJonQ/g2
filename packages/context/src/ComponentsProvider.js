export { ContextSystemProvider as ComponentsProvider } from './ContextSystemProvider';
