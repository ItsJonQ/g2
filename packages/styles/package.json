{
    "name": "@wp-g2/styles",
    "version": "0.0.47",
    "description": "Style system for G2 Components.",
    "main": "dist/cjs/index.js",
    "module": "dist/es/index.js",
    "sideEffects": false,
    "types": "types",
    "scripts": {
        "clean": "rm -rf dist",
        "build:es": "BUILD_FORMAT=es zero build -d dist/es",
        "build:cjs": "BUILD_FORMAT=cjs zero build -d dist/cjs",
        "build": "npm run build:cjs && npm run build:es -- --no-clean",
        "test": "echo \"Error: no test specified\" && exit 1"
    },
    "author": "Jon Quach <hello@jonquach.com> (https://jonquach.com)",
    "license": "MIT",
    "dependencies": {
        "@emotion/core": "^10.0.28",
        "@emotion/is-prop-valid": "^0.8.8",
        "@wp-g2/utils": "^0.0.47",
        "create-emotion": "^10.0.27",
        "cssjanus": "^1.3.2",
        "dash-get": "^1.0.2",
        "emotion": "^10.0.27",
        "emotion-theming": "^10.0.27",
        "styled-griddie": "^0.1.3",
        "stylis-plugin-css-variables": "^1.0.4",
        "stylis-plugin-rtl": "^1.0.0"
    },
    "peerDependencies": {
        "react": "^16.8.0",
        "react-dom": "^16.8.0"
    }
}
