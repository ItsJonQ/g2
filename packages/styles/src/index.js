export * from './core';
export { compiler } from './system';
export { cache, css, cx, injectGlobal, keyframes } from './style-system';

export * from './styled';
export * from './hooks';
export * from './mixins';
export * from './namespaces';
export {
	config,
	createTheme,
	getIsReducedMotion,
	SUPPORTED_COLORS,
} from './theme';

export * from './presets/ui';

export * from './components';
