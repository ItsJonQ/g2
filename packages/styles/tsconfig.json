{
	"extends": "../../tsconfig.base.json",
	"compilerOptions": {
		"rootDir": "src",
		"declarationDir": "types"
	},
	"include": ["src/**/*"],
	"references": [{ "path": "../create-styles" }, { "path": "../utils" }]
}
