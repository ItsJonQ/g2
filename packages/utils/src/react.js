import { Children, isValidElement } from 'react';

/**
 * Merges React `ref` together.
 */
export { default as mergeRefs } from 'react-merge-refs';

/**
 * Copies non-react specific statics from a child component to a parent component.
 * Similar to Object.assign, but with React static keywords prevented from being overridden.
 */
export { default as hoistNonReactStatics } from 'hoist-non-react-statics';

/**
 * Gets a collection of available children elements from a React component's children prop.
 *
 * @param {import('react').ReactNode} children
 *
 * @return {import('react').ReactNodeArray} An array of available children.
 */
export function getValidChildren(children) {
	if (typeof children === 'string') return [children];

	return Children.toArray(children).filter((child) => isValidElement(child));
}

/**
 * Gets the displayName of a React Component or element.
 *
 * @param {string | import('react').ComponentType} tagName
 *
 * @return {string} The display name of the Component / tagName.
 */
export function getDisplayName(tagName) {
	const displayName =
		typeof tagName === 'string'
			? tagName
			: tagName.displayName || tagName.name || 'Component';

	return displayName;
}

/**
 * Checks to see if a `children` prop is a render function prop.
 *
 * @param {*} children
 *
 * @return {boolean} True, if children is a render function prop.
 */
export function isRenderProp(children) {
	return typeof children === 'function';
}

/**
 * Handles the rendering of a React component's children prop, which
 * may be a render function prop.
 *
 * @param {*} children Children to render.
 * @param {*} props Props to pass into a (potential) children render function.
 *
 * @return {*} The rendered children.
 */
export function renderChildren(children, props = {}) {
	if (isRenderProp(children)) {
		// eslint-disable-next-line no-unused-vars
		const { children: _, ...rest } = props;
		return children(rest);
	}

	return children;
}
