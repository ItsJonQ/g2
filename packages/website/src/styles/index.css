body {
  padding-top: 0;
}
