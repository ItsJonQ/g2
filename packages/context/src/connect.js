import { css, cx } from '@wp-g2/styles';
import { hoistNonReactStatics, is, kebabCase, uniq } from '@wp-g2/utils';
import React, { forwardRef } from 'react';

import { useComponentsContext } from './ComponentsProvider';
import { cns, ns } from './utils';

const REACT_TYPEOF_KEY = '$$typeof';

/**
 * Special key where the connected namespaces are stored.
 * This is attached to Context connected components as a static property.
 */
const CONNECT_NAMESPACE = '__wpComponentsKey__';

/**
 * Connects a component to the G2 Context system using a higher-order component.
 * The workflow for this connect HOC is similar to that of Redux.
 *
 * @example
 * ```jsx
 * const HelloComponent = () => <div>Hello</div>
 * const ConnectedHello = connect(HelloComponent, 'Hello')
 * ```
 *
 * @param {React.Component} Component The component to connect.
 * @param {Array<string>|string} namespace The name for the component to be identified by the Context system.
 * @param {object} options Options to modify the connected component.
 * @returns {React.Component} The connected component.
 */
export function connect(Component, namespace, options = {}) {
	const { pure = true } = options;
	const displayName = is.array(namespace)
		? namespace[0]
		: namespace || Component.name;
	const key = namespace || Component.name;

	const render = ({ className, ...props }, forwardedRef) => {
		// eslint-disable-next-line
		const context = useComponentsContext();
		let contextProps;

		/**
		 * It's possible to connect (register) a component under multiple namespaces.
		 * This is done when passing the namespace as an array (Array<string>).
		 *
		 * To properly retrieve props from the Context system, we must accomodate
		 * both singular and multi connections.
		 */
		if (is.array(key)) {
			contextProps = key.reduce((acc, k) => {
				const v = context[k];
				return is.plainObject(v) ? { ...acc, ...v } : acc;
			}, {});
		} else {
			contextProps =
				context[key] ||
				// Fallback
				{};
		}

		const {
			_overrides: overrideProps = {},
			css: contextCSS,
			...otherContextProps
		} = contextProps;

		const initialMergedProps = is.plainObject(contextProps)
			? { ...otherContextProps, ...props }
			: props;

		const { children, renderChildren, ...mergedProps } = initialMergedProps;

		const classes = cx(
			// Resolve custom CSS from ComponentsProvider
			css(contextCSS),
			getStyledClassNameFromKey(key),
			className,
		);

		// Provides the ability to customize the render of the component.
		const rendered = is.function(renderChildren)
			? renderChildren({ children, ...mergedProps })
			: children;

		return (
			<Component
				{...cns()}
				{...ns(displayName)}
				{...mergedProps}
				{...overrideProps}
				className={classes}
				forwardedRef={forwardedRef}
			>
				{rendered}
			</Component>
		);
	};

	const ForwardedComponent = forwardRef(render);
	const ConnectedComponent = pure
		? React.memo(ForwardedComponent)
		: ForwardedComponent;

	let mergedNamespace = Component[CONNECT_NAMESPACE] || [displayName];

	/**
	 * Consolidate (merge) namespaces before attaching it to the component.
	 */
	if (is.array(namespace)) {
		mergedNamespace = [...mergedNamespace, ...namespace];
	}
	if (is.string(namespace)) {
		mergedNamespace = [...mergedNamespace, namespace];
	}

	ConnectedComponent.displayName = displayName;
	ConnectedComponent[CONNECT_NAMESPACE] = uniq(mergedNamespace);

	return hoistNonReactStatics(ConnectedComponent, Component);
}

/**
 * Generates the connected component CSS className based on the namespace.
 *
 * @param {string} displayName The name of the connected component.
 * @returns {string} The generated CSS className.
 */
function getStyledClassName(displayName) {
	if (!displayName || !is.string(displayName)) return '';

	return `wp-components-${kebabCase(displayName)}`;
}

/**
 * Generates the connected component CSS className based on the namespace.
 *
 * @param {string} displayName The name of the connected component.
 * @returns {string} The generated CSS className.
 */
function getStyledClassNameFromKey(key) {
	if (!key) return '';

	if (is.array(key)) {
		return cx(uniq(key).map(getStyledClassName));
	}
	if (is.string(key)) {
		return getStyledClassName(key);
	}

	return '';
}

/**
 * Attempts to retrieve the connected namespace from a component.
 *
 * @param {React.Component} Component The component to retrieve a namespace from.
 * @returns {Array<string>} The connected namespaces.
 */
export function getConnectNamespace(Component) {
	return (Component && Component[CONNECT_NAMESPACE]) || [];
}

/**
 * Checks to see if a component is connected within the Context system.
 *
 * @param {React.Component} Component The component to retrieve a namespace from.
 * @param {Array<string>|string} match The namespace to check.
 * @returns {boolean} The result.
 */
export function hasConnectNamespace(Component, match) {
	const BaseComponent = Component[REACT_TYPEOF_KEY]
		? Component.type
		: Component;

	if (is.string(match)) {
		return getConnectNamespace(BaseComponent).includes(match);
	}
	if (is.array(match)) {
		return match.some((result) =>
			getConnectNamespace(BaseComponent).includes(result),
		);
	}

	return false;
}

export const getNamespace = getConnectNamespace;
export const hasNamespace = hasConnectNamespace;
