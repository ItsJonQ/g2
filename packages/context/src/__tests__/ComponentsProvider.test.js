import { render } from '@testing-library/react';
import { View } from '@wp-g2/styles';
import React from 'react';

import { ComponentsProvider } from '../ComponentsProvider';
import { connect } from '../connect';

describe('props', () => {
	test('should render correctly', () => {
		const Olaf = (props) => <View {...props} />;
		const ConnectedOlaf = connect(Olaf, 'Olaf');
		const { container } = render(
			<ComponentsProvider>
				<ConnectedOlaf />
			</ComponentsProvider>,
		);

		expect(container.firstChild).toMatchSnapshot();
	});

	test('should render context props', () => {
		const Olaf = ({ quote, ...props }) => <View {...props}>{quote}</View>;
		const ConnectedOlaf = connect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				quote: 'Warm Hugs',
			},
		};

		const { container } = render(
			<ComponentsProvider value={contextValue}>
				<ConnectedOlaf />
			</ComponentsProvider>,
		);

		expect(container.firstChild).toMatchSnapshot();
		expect(container.firstChild.innerHTML).toContain('Warm Hugs');
	});

	test('should render css props', () => {
		const Olaf = ({ quote, ...props }) => <View {...props}>{quote}</View>;
		const ConnectedOlaf = connect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				css: [{ background: 'white' }, `padding: 20px`],
			},
		};

		const { container } = render(
			<>
				<ComponentsProvider value={contextValue}>
					<ConnectedOlaf className="olaf" />
				</ComponentsProvider>
				<ConnectedOlaf className="olaf" />
			</>,
		);

		const [first, second] = [...container.querySelectorAll('.olaf')];

		expect(container.firstChild).toMatchSnapshot();

		const firstCSS = first.classList.toString();
		const secondCSS = second.classList.toString();
		expect(firstCSS).not.toEqual(secondCSS);
	});

	test('should render css props + provided css', () => {
		const Olaf = ({ quote, ...props }) => <View {...props}>{quote}</View>;
		const ConnectedOlaf = connect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				css: [{ background: 'white' }, { border: '2px solid blue' }],
			},
		};

		const { container } = render(
			<>
				<ComponentsProvider value={contextValue}>
					<ConnectedOlaf
						className="olaf"
						css={`
							font-weight: bold;
						`}
					/>
				</ComponentsProvider>
			</>,
		);

		expect(container.firstChild).toMatchSnapshot();

		const el = container.querySelector('.olaf');

		expect(el).toHaveStyle(`background: white`);
		expect(el).toHaveStyle(`font-weight: bold`);
	});

	test('should render _override props', () => {
		const Olaf = ({ quote, ...props }) => <View {...props}>{quote}</View>;
		const ConnectedOlaf = connect(Olaf, 'Olaf');

		const contextValue = {
			Olaf: {
				_overrides: {
					quote: 'Warm Hugs!',
				},
			},
		};

		const { container } = render(
			<>
				<ComponentsProvider value={contextValue}>
					<ConnectedOlaf className="olaf" quote="Hello" />
				</ComponentsProvider>
			</>,
		);

		expect(container.firstChild).toMatchSnapshot();

		const el = container.querySelector('.olaf');

		expect(el.innerHTML).toContain('Warm Hugs!');
		expect(el.innerHTML).not.toContain('Hello');
	});
});
