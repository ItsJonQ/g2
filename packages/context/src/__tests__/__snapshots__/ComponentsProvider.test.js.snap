// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`props should render _override props 1`] = `
<div
  data-system-theme-provider="true"
>
  <div
    class="wp-components-olaf olaf css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  >
    Warm Hugs!
  </div>
</div>
`;

exports[`props should render context props 1`] = `
<div
  data-system-theme-provider="true"
>
  <div
    class="wp-components-olaf css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  >
    Warm Hugs
  </div>
</div>
`;

exports[`props should render correctly 1`] = `
<div
  data-system-theme-provider="true"
>
  <div
    class="wp-components-olaf css-j1yu9m"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  />
</div>
`;

exports[`props should render css props + provided css 1`] = `
<div
  data-system-theme-provider="true"
>
  <div
    class="wp-components-olaf olaf css-1r6lrag"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  />
</div>
`;

exports[`props should render css props 1`] = `
<div
  data-system-theme-provider="true"
>
  <div
    class="wp-components-olaf olaf css-1dqvrac"
    data-g2-c16t="true"
    data-g2-component="Olaf"
  />
</div>
`;
